orttraining/orttraining/training_api/training_session.cc

```diff
@@ -7,19 +7,26 @@ namespace onnxruntime {
 namespace training {
 namespace api {
 
-TrainingSession::TrainingSession(const Environment& session_env,
+TrainingSession::TrainingSession(Environment& session_env,
                                  const SessionOptions& session_options,
                                  const std::vector<std::shared_ptr<IExecutionProvider>>& providers,
                                  CheckpointState* state,
                                  const ModelIdentifiers& model_identifiers)
-    : state_{state},
-      module_{std::make_unique<Module>(model_identifiers.train_model, state_,
-                                       session_options, session_env, providers, model_identifiers.eval_model)},
-      optimizer_{model_identifiers.optim_model.has_value()
-                     ? std::make_unique<Optimizer>(
-                           model_identifiers.optim_model.value(), state_,
-                           session_options, session_env, providers)
-                     : std::unique_ptr<Optimizer>()} {}
+    : state_{state} {
+  for (auto& provider : providers) {
```

Comment:
Several points we need to review and discuss:
1. Shall we register all EP's allocators in environment or just cpu & cuda EP?
2. Shall we register cuda's pinned allocator or not?
3. I assume there is no need to unregister allocator, right? Just let it release automatically when this TrainingSession object dies.
4. As the parameter session_options is const, It is the caller (of the API CreateTrainingSession)'s responsibility to set session_options.AddConfigEntry(kOrtSessionOptionsConfigUseEnvAllocators, "1") (which doesn't make sense to me), or we change the API's signature, to remove the const qualifier so we can add this config entry inside this function?

Reply:
For #4, I think we can just create a local session options object inside this function and do the change accordingly on the local object

Reply:
On-Device training only supports CPU and CUDA EPs. So yes let's register all EPs allocators. Given the environment will own the allocators (because we are sharing them across the sessions) then wont they be released once all the sessions and then the env is destroyed. Do we need an explicit unregister allocator call?